idaes/apps/grid_integration/bidder.py

```diff
@@ -0,0 +1,504 @@
+                    continue
+                elif power in bids[t][gen]:
+                    bids[t][gen][power] = min(bids[t][gen][power], marginal_cost)
+                else:
+                    bids[t][gen][power] = marginal_cost
+
+        for t in time_index:
+
+            # make sure the orignal points in the bids
+            for power, marginal_cost in self.bidding_model_object.default_bids.items():
+                if power not in bids[t][gen]:
+                    bids[t][gen][power] = marginal_cost
+
+            pmin = self.bidding_model_object.pmin
+
+            # sort the curves by power
+            bids[t][gen] = dict(sorted(bids[t][gen].items()))
+
+            # make sure the curve is nondecreasing
+            pre_power = pmin
+            for power, marginal_cost in bids[t][gen].items():
+
+                # ignore pmin, because min load cost is special
+                if pre_power == pmin:
+                    pre_power = power
+                    continue
+                bids[t][gen][power] = max(bids[t][gen][power], bids[t][gen][pre_power])
+                pre_power = power
+
+            # calculate the actual cost
+            pre_power = 0
+            pre_cost = 0
+            for power, marginal_cost in bids[t][gen].items():
+
+                delta_p = power - pre_power
+                bids[t][gen][power] = pre_cost + marginal_cost * delta_p
+                pre_power = power
+                pre_cost += marginal_cost * delta_p
+
+        # check if bids are convex
+        for t in bids:
+            for gen in bids[t]:
+                if not self._is_convex_bid(bids[t][gen]):
+                    raise RuntimeError(
+                        f"Bids for generator {gen} at hour {t} is not convex!"
+                    )
+
+        return bids
+
+    @staticmethod
+    def _is_convex_bid(bids):
+
+        """
+        This method checks the convexity of a bid at a single time period from a
+         single generator.
+
+        Arguments:
+            bids: a bids at at a single time period from a single generator,
+            which is a dictionary whose keys are the power outputs and the values
+             are the corresponding costs. {power: cost}
+
+        Returns:
+            bids: the bid we computed. It is a dictionary that has this structure. {t: {gen:{power: cost}}}.
+        """
+
+        power = list(bids.keys())
+        power.sort()
+
+        idx = 0
+        delta_p = []
+        marginal_cost = []
+
+        # calculate marginal costs (slope)
+        while idx < len(power) - 1:
+            delta_p.append(power[idx + 1] - power[idx])
+            marginal_cost.append(
+                (bids[power[idx + 1]] - bids[power[idx]]) / delta_p[-1]
+            )
+            idx += 1
+
+        # check whether the marginal costs are sorted <=> convex
+        idx = 0
+        while idx < len(marginal_cost) - 1:
+            if marginal_cost[idx] > marginal_cost[idx + 1]:
```

Comment:
This check can fail due to numerical comparison. This was causing issues for me when the marginal cost was supposed to be constant, but the the true values were different by 1e-12. We should round within precision for this check.